pkg/relayer/session/sessiontree.go

```diff
@@ -0,0 +1,192 @@
+package session
+
+import (
+	"crypto/sha256"
+	"os"
+	"path/filepath"
+	sessiontypes "pocket/x/session/types"
+	"sync"
+
+	"github.com/pokt-network/smt"
+)
+
+var _ SessionTree = (*sessionTree)(nil)
+
+// sessionTree is an implementation of the SessionTree interface.
+type sessionTree struct {
+	// sessionMu is a mutex used to protect sessionTree operations from concurrent access.
+	sessionMu *sync.Mutex
+
+	// session is the Session corresponding to the SMST (Sparse Merkle State Tree).
+	session *sessiontypes.Session
+
+	// tree is the SMST (Sparse Merkle State Tree) corresponding the session.
+	tree *smt.SMST
+
+	// claimedRoot is the root hash of the SMST needed for submitting the claim.
+	// If it holds a non-nil value, it means that the SMST has been flushed, committed to disk
+	// and no more updates can be made to it. It also indicates that a proof could be generated
+	// using ProveClosest function.
+	claimedRoot []byte
+
+	// proofPath is the path for which the proof was generated.
+	proofPath []byte
+
+	// proof is the generated proof for the session given a proofPath.
+	proof *smt.SparseMerkleClosestProof
+
+	// treeStore is the KVStore used to store the SMST.
+	treeStore smt.KVStore
+
+	// storePath is the path to the KVStore used to store the SMST.
+	// It is created from the storePrefix and the session.sessionId.
+	// We keep track of it so we can use it at the end of the claim/proof lifecycle
+	// to delete the KVStore when it is no longer needed.
+	storePath string
+
+	// removeFromRelayerSessions is a function that removes the sessionTree from the RelayerSessionsManager.
+	// Since the sessionTree has no knowledge of the RelayerSessionsManager, we pass this callback
+	// from the session manager to the sessionTree so it can remove itself from the RelayerSessionsManager
+	// when it is no longer needed.
+	removeFromRelayerSessions func(session *sessiontypes.Session)
+}
+
+// NewSessionTree creates a new sessionTree from a Session and a storePrefix. It also takes a function
+// removeFromRelayerSessions that removes the sessionTree from the RelayerSessionsManager.
+// It returns an error if the KVStore fails to be created.
+func NewSessionTree(
+	session *sessiontypes.Session,
+	storesDirectory string,
+	removeFromRelayerSessions func(session *sessiontypes.Session),
+) (SessionTree, error) {
+	// join the storePrefix and the session.sessionId to create a unique storePath
+	storePath := filepath.Join(storesDirectory, session.SessionId)
+
+	treeStore, err := smt.NewKVStore(storePath)
```

Comment:
```suggestion
	// NOTE: If storePath is a prexisting path it will be used to open a KVStore there. This will result in unexpected behaviour
	treeStore, err := smt.NewKVStore(storePath)
```